sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs

```diff
@@ -1054,5 +1056,291 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
+            }
+
+            async Task<Page<QueryResultPage>> FirstPageFunc(int? pageSizeHint)
+            {
+                using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(QueryEvents)}");
+                scope.Start();
+                try
+                {
+                    Response<QueryResultPage> response = await _queryRestClient
+                        .ExecuteAsync(queryRequest, options?.StoreType?.ToString(), null, null, cancellationToken)
+                        .ConfigureAwait(false);
+
+                    var frame = new QueryResultPage[]
+                    {
+                            response.Value
+                    };
+
+                    return Page.FromValues(frame, response.Value.ContinuationToken, response.GetRawResponse());
+                }
+                catch (Exception ex)
+                {
+                    scope.Failed(ex);
+                    throw;
+                }
+            }
+
+            async Task<Page<QueryResultPage>> NextPageFunc(string nextLink, int? pageSizeHint)
+            {
+                using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(QueryEvents)}");
+                scope.Start();
+                try
+                {
+                    Response<QueryResultPage> response = await _queryRestClient
+                        .ExecuteAsync(queryRequest, options?.StoreType?.ToString(), nextLink, null, cancellationToken)
+                        .ConfigureAwait(false);
+
+                    var frame = new QueryResultPage[]
+                    {
+                            response.Value
+                    };
+
+                    return Page.FromValues(frame, response.Value.ContinuationToken, response.GetRawResponse());
+                }
+                catch (Exception ex)
+                {
+                    scope.Failed(ex);
+                    throw;
+                }
+            }
+
+            return PageableHelpers.CreateAsyncEnumerable(FirstPageFunc, NextPageFunc);
+        }
+
+        private Pageable<QueryResultPage> QueryEventsInternal(
+            TimeSeriesId timeSeriesId,
+            DateTimeOffset startTime,
+            DateTimeOffset endTime,
+            QueryEventsRequestOptions options,
+            CancellationToken cancellationToken)
+        {
+            var searchSpan = new DateTimeRange(startTime, endTime);
+            var queryRequest = new QueryRequest
+            {
+                GetEvents = new GetEvents(timeSeriesId, searchSpan)
+            };
+
+            if (options != null)
+            {
+                if (options.Filter != null)
+                {
+                    queryRequest.GetEvents.Filter = new TimeSeriesExpression(options.Filter);
+                }
+
+                if (options.ProjectedProperties != null)
```

Comment:
Is this common code between sync and async?

Reply:
Yup. Already factored out.